pkg/pipeline/write/write_loki.go

```diff
@@ -0,0 +1,252 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package write
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"github.com/netobserv/flowlogs2metrics/pkg/api"
+	"math"
+	"strings"
+	"time"
+
+	"github.com/netobserv/flowlogs2metrics/pkg/config"
+
+	logAdapter "github.com/go-kit/kit/log/logrus"
+	jsonIter "github.com/json-iterator/go"
+	"github.com/netobserv/loki-client-go/loki"
+	"github.com/netobserv/loki-client-go/pkg/backoff"
+	"github.com/netobserv/loki-client-go/pkg/urlutil"
+	"github.com/prometheus/common/model"
+	log "github.com/sirupsen/logrus"
+)
+
+var (
+	keyReplacer = strings.NewReplacer("/", "_", ".", "_", "-", "_")
+)
+
+type emitter interface {
+	Handle(labels model.LabelSet, timestamp time.Time, record string) error
+}
+
+// Loki record writer
+type Loki struct {
+	lokiConfig loki.Config
+	apiConfig  api.WriteLoki
+	client     emitter
+	timeNow    func() time.Time
+	ready      bool
+}
+
+func (l *Loki) IsReady() bool {
+	return l.ready
+}
+
+func buildLokiConfig(c *api.WriteLoki) (loki.Config, error) {
+	batchWait, err := time.ParseDuration(c.BatchWait)
+	if err != nil {
+		return loki.Config{}, fmt.Errorf("failed in parsing BatchWait : %v", err)
+	}
+
+	timeout, err := time.ParseDuration(c.Timeout)
+	if err != nil {
+		return loki.Config{}, fmt.Errorf("failed in parsing Timeout : %v", err)
+	}
+
+	minBackoff, err := time.ParseDuration(c.MinBackoff)
+	if err != nil {
+		return loki.Config{}, fmt.Errorf("failed in parsing MinBackoff : %v", err)
+	}
+
+	maxBackoff, err := time.ParseDuration(c.MaxBackoff)
+	if err != nil {
+		return loki.Config{}, fmt.Errorf("failed in parsing MaxBackoff : %v", err)
+	}
+
+	cfg := loki.Config{
+		TenantID:  c.TenantID,
+		BatchWait: batchWait,
+		BatchSize: c.BatchSize,
+		Timeout:   timeout,
+		BackoffConfig: backoff.BackoffConfig{
+			MinBackoff: minBackoff,
+			MaxBackoff: maxBackoff,
+			MaxRetries: c.MaxRetries,
+		},
+		Client: c.ClientConfig,
+	}
+	var clientURL urlutil.URLValue
+	err = clientURL.Set(strings.TrimSuffix(c.URL, "/") + "/loki/api/v1/push")
+	if err != nil {
+		return cfg, fmt.Errorf("failed to parse client URL: %w", err)
+	}
+	cfg.URL = clientURL
+	return cfg, nil
+}
+
+func (l *Loki) ProcessRecord(record config.GenericMap) error {
+	if !l.IsReady() {
+		return errors.New("WriteLoki is not ready")
+	}
+
+	// Get timestamp from record (default: TimeFlowStart)
+	timestamp := l.extractTimestamp(record)
+
+	labels := model.LabelSet{}
+
+	// Add static labels from config
+	for k, v := range l.apiConfig.StaticLabels {
+		labels[k] = v
+	}
+
+	l.addNonStaticLabels(record, labels)
+
+	// Remove labels and configured ignore list from record
+	ignoreList := append(l.apiConfig.IgnoreList, l.apiConfig.Labels...)
+	for _, label := range ignoreList {
+		delete(record, label)
```

Comment:
Just a remark here: the input map is mutated. In goflow-kube that was ok because the Loki write was a terminal action. Maybe it's the same here. But I prefer to catch your attention here. Also, depending on how we implement the pipeline fork mechanism later, it may become an issue.

Reply:
@jotak with the new version of the PR (that sends the records over a channel to Loki writer) I think that this is ok. This record is now a copy of the original one and there is no influence on the continuation of the pipeline so this should be ok.

Thanks for raising that :-)